### PR TITLE
New tpm_quote crate to attest IMA with TPM

### PR DESCRIPTION
**Pull request by grydz**

### CODE REVIEW

File: tpm_quote/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "tpm_quote"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+# ecdsa = { version = "0.16", features = ["sha2", "verifying"] }
+hex = "0.4"
+p256 = { version = "0.13", features = ["arithmetic", "pkcs8", "ecdsa-core"] }
```
[the remark this answers is not in the document]
Comment (grydz) [reply tpm_quote/Cargo.toml]: See 95ed88ec85c188e5ae62417569f8f25285a84cd6.

File: tpm_quote/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    //     assert!(ek_pub.marshall().is_ok());
+    //     assert!(ak_pub.marshall().is_ok());
+
+    //     let (attestation_data, signature) = attest_pcr_10(&mut context, &ak_handle, None).unwrap();
+
+    //     let nonce = verify_quote_signature(attestation_data, signature, ak_pub).unwrap();
+
+    //     assert_eq!(nonce, vec![0xff; 16]);
+    // }
```
[the remark this answers is not in the document]
Comment (grydz) [reply tpm_quote/src/lib.rs]: For future use when we'll have vTPM in the CI.

File: tpm_quote/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    --message=pcr_quote.plain \
+    --signature=pcr_quote.sig \
+    --hash-algorithm=sha256
+```
+
+Set env variable depending on your TPM device
+
+```console
+export TPM2TOOLS_TCTI="device:/dev/tpm0"
+```
```
Comment (heavenboy8) [inline tpm_quote/README.md:49]: Mention `/dev/tpm0r0` in GCP?

File: tpm_quote/src/command.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    },
+    structures::{
+        Attest, AttestInfo, Data, HashScheme, PcrSelectionList, Signature, SignatureScheme,
+    },
+    Context,
+};
+
+use std::convert::TryFrom;
+
+pub(crate) fn attest_pcr(
```
Comment (heavenboy8) [inline tpm_quote/src/command.rs:16]: Is it possible to add a comment describing all checks made during the attestation?

File: tpm_quote/src/command.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                "Attested did not contain the expected variant.".to_owned(),
+            ));
+        }
+    }
+
+    Ok((attestation, signature))
+}
+
+#[allow(dead_code)]
+pub(crate) fn certify_key_handle(
```
Comment (heavenboy8) [inline tpm_quote/src/command.rs:68]: Why dead code?

File: tpm_quote/src/convert.rs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    let mut point = [0x00u8; 65];
+    point[0] = 0x04;
+    point[1..33].copy_from_slice(&x[..]);
+    point[33..].copy_from_slice(&y[..]);
+
+    // let point =
+    //     EncodedPoint::from_bytes(point).map_err(|e| Error::VerificationError(format!("{e}")))?;
+    // let pk = p256::PublicKey::from_encoded_point(&point).unwrap();
+
```
Comment (heavenboy8) [inline tpm_quote/src/convert.rs:72]: Remove?

File: tpm_quote/src/convert.rs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    let pk = p256::PublicKey::from_sec1_bytes(&point).unwrap();
```
Comment (heavenboy8) [inline tpm_quote/src/convert.rs:73]: don't use unwrap?

File: tpm_quote/src/key.rs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+) -> Result<(KeyHandle, Public), Error> {
+    let tpm_handle = TpmHandle::Persistent(PersistentTpmHandle::new(index)?);
+    let key_handle: KeyHandle = context.tr_from_tpm_public(tpm_handle)?.into();
+
+    let (public, _, _) = context.read_public(key_handle)?;
+
+    Ok((key_handle, public))
+}
+#[allow(dead_code)]
+pub(crate) fn get_content_from_nv_handle(
```
Comment (heavenboy8) [inline tpm_quote/src/key.rs:26]:
```suggestion
}

#[allow(dead_code)]
pub(crate) fn get_content_from_nv_handle(
```

File: tpm_quote/src/key.rs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    index: u32,
+) -> Result<(KeyHandle, Public), Error> {
+    let tpm_handle = TpmHandle::Persistent(PersistentTpmHandle::new(index)?);
+    let key_handle: KeyHandle = context.tr_from_tpm_public(tpm_handle)?.into();
+
+    let (public, _, _) = context.read_public(key_handle)?;
+
+    Ok((key_handle, public))
+}
+#[allow(dead_code)]
```
Comment (heavenboy8) [inline tpm_quote/src/key.rs:25]: why dead code?

File: tpm_quote/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+#[allow(clippy::type_complexity)]
+pub fn get_quote(
+    context: &mut Context,
+    pcr_list: &[u8],
+    nonce: Option<&[u8]>,
+) -> Result<(Vec<u8>, Vec<u8>, Vec<u8>), Error> {
+    let pcr_list = pcr_list
+        .iter()
+        .map(|n| (1 << (*n as u32)).try_into().unwrap())
```
Comment (heavenboy8) [inline tpm_quote/src/lib.rs:28]: Don't unwrap?

File: tpm_quote/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+pub mod error;
+pub mod key;
+pub mod verify;
+
+#[allow(clippy::type_complexity)]
+pub fn get_quote(
+    context: &mut Context,
+    pcr_list: &[u8],
+    nonce: Option<&[u8]>,
+) -> Result<(Vec<u8>, Vec<u8>, Vec<u8>), Error> {
```
Comment (heavenboy8) [inline tpm_quote/src/lib.rs:25]: Add a comment to help understanding the return type?

File: tpm_quote/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+mod tests {
+    use crate::{get_quote, verify_quote};
+    use tss_esapi::{tcti_ldr::TctiNameConf, Context};
+
+    use test_log::test;
+
+    #[test]
+    #[ignore]
+    fn test_get_quote() {
+        let tcti = TctiNameConf::from_environment_variable().expect("TCTI not found");
```
Comment (heavenboy8) [inline tpm_quote/src/lib.rs:90]: May be we should add the env variable in the CI yaml ?

File: tpm_quote/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+#[cfg(test)]
+mod tests {
+    use crate::{get_quote, verify_quote};
+    use tss_esapi::{tcti_ldr::TctiNameConf, Context};
+
+    use test_log::test;
+
+    #[test]
+    #[ignore]
```
Comment (heavenboy8) [inline tpm_quote/src/lib.rs:88]: Could you not ignore the test but make a check in the function. If TPM is enable then return the result of the assert, if not return true. 

Example here: https://github.com/Cosmian/tee-tools/blob/1717d380ce8065c46501a18ad22a982c5ca23362/sev_quote/src/quote.rs#L184

I will create a specific runner for that

File: tpm_quote/src/verify.rs [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+            );
+
+            Ok((x, y))
+        }
+        _ => Err(Error::VerificationError(
+            "unsupported public key".to_owned(),
+        )),
+    }?;
+
+    let mut point = [0x00u8; 65];
```
Comment (heavenboy8) [inline tpm_quote/src/verify.rs:102]: Create a const variable for `65` ?

File: tpm_quote/src/verify.rs [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+    point[33..].copy_from_slice(&y[..]);
+
+    let point =
+        EncodedPoint::from_bytes(point).map_err(|e| Error::VerificationError(format!("{e}")))?;
+    let point =
+        AffinePoint::<NistP256>::from_encoded_point(&point).expect("public key not on curve P-256");
+
+    let verifier = VerifyingKey::from_affine(point)?;
+    let signature = p256::ecdsa::Signature::from_scalars(r, s)?;
+    verifier.verify(&message[..], &signature).unwrap();
```
Comment (heavenboy8) [inline tpm_quote/src/verify.rs:114]: do not unwrap?

File: tpm_quote/src/verify.rs [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+
+    let mut point = [0x00u8; 65];
+    point[0] = 0x04;
+    point[1..33].copy_from_slice(&x[..]);
+    point[33..].copy_from_slice(&y[..]);
+
+    let point =
+        EncodedPoint::from_bytes(point).map_err(|e| Error::VerificationError(format!("{e}")))?;
+    let point =
+        AffinePoint::<NistP256>::from_encoded_point(&point).expect("public key not on curve P-256");
```
Comment (heavenboy8) [inline tpm_quote/src/verify.rs:110]: do not expect?

File: tpm_quote/src/command.rs [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+        (
+            Some(AuthSession::Password),
+            Some(AuthSession::Password),
+            None,
+        ),
+        |ctx| {
+            ctx.certify(
+                (*key_handle).into(),
+                *sign_key_handle,
+                Data::try_from(qualifying_data).unwrap(),
```
Comment (heavenboy8) [inline tpm_quote/src/command.rs:90]: Do not unwrap?

File: tpm_quote/src/convert.rs (hunk 6 above)
Comment (grydz) [reply tpm_quote/src/convert.rs]: Done in c00096d152de059ca8b52dcb2b2d3c185ca9ed03.

File: tpm_quote/src/convert.rs (hunk 7 above)
Comment (grydz) [reply tpm_quote/src/convert.rs]: Done in c00096d152de059ca8b52dcb2b2d3c185ca9ed03.

File: tpm_quote/src/lib.rs (hunk 13 above)
Comment (grydz) [reply tpm_quote/src/lib.rs]: Done in c00096d152de059ca8b52dcb2b2d3c185ca9ed03.

File: tpm_quote/src/verify.rs (hunk 14 above)
Comment (grydz) [reply tpm_quote/src/verify.rs]: Removed in c00096d152de059ca8b52dcb2b2d3c185ca9ed03.

File: tpm_quote/src/verify.rs (hunk 16 above)
Comment (grydz) [reply tpm_quote/src/verify.rs]: Done in c00096d152de059ca8b52dcb2b2d3c185ca9ed03.

File: tpm_quote/src/verify.rs (hunk 15 above)
Comment (grydz) [reply tpm_quote/src/verify.rs]: Done in 4c89b1e036c2e9819789901d7ffb13bb0212a650.

File: tpm_quote/src/command.rs (hunk 17 above)
Comment (grydz) [reply tpm_quote/src/command.rs]: See 4c89b1e036c2e9819789901d7ffb13bb0212a650.

File: tpm_quote/src/command.rs (hunk 4 above)
Comment (grydz) [reply tpm_quote/src/command.rs:12]: Added comments on public API `get_quote()` and `verify_quote()`.

File: tpm_quote/src/lib.rs (hunk 11 above)
Comment (grydz) [reply tpm_quote/src/lib.rs:30]: See 44184e424d7c2f74b6cb058a6ac77438c4651445.

File: tpm_quote/src/key.rs (hunk 8 above)
Comment (grydz) [reply tpm_quote/src/key.rs]: See 44184e424d7c2f74b6cb058a6ac77438c4651445.

File: tpm_quote/src/lib.rs (hunk 10 above)
Comment (grydz) [reply tpm_quote/src/lib.rs]: See 44184e424d7c2f74b6cb058a6ac77438c4651445.

**Review by heavenboy8 — APPROVED** [review]

**Pull request merged**